queries/racket/highlights.scm

```diff
@@ -124,18 +120,25 @@
 ;;                         Solve conflicts                          ;;
 ;;------------------------------------------------------------------;;
 
-;; See `:h treesitter`, and search `priority`
+;; See `:h treesitter-highlight-priority`
 
 (list
  .
  (symbol) @include
  (#eq? @include "require")
  (#set! "priority" 101))
 
-(quote
-  .
-  (symbol)
-  (#set! "priority" 105)) @symbol
+;; quote
+
+(quote (#set! "priority" 105)) @symbol
+
+(list
+ .
+ (symbol) @_f
+ (#eq? @_f "quote")
+ (#set! "priority" 105)) @symbol
```

Comment:
Is it necessary to set a priority to these captures (and captures after it). Can't it be moved down instead?

Reply:
Yes. It's here for nested structure.

For example, It would be impossible to highlight this whole sexp as the same color without priority

```racket
(quote (list a b c (list d e f)))
```

Reply:
I'm still a bit not sold on the need for priorities, as captures applied earlier already has lower precedence

Reply:
> captures applied earlier already has lower precedence

precedence != priority.

from the docs `:h treesitter-highlight-priority`

> Treesitter uses |nvim_buf_set_extmark()| to set highlights with a default priority of 100

earlier query don't have lower priority.

Another reason is it works with priority, and don't work without priority.

To make sure we are talking about the same thing. The query without priority is:

```scheme
(list
 .
 (symbol) @_f
 (#eq? @_f "quote")) @symbol
```

A picture without priority:
![Screenshot_20230623_190120](https://github.com/nvim-treesitter/nvim-treesitter/assets/39000776/8a0106fb-2c48-4064-8209-48ab186e4e41)

with priority:

![Screenshot_20230623_190209](https://github.com/nvim-treesitter/nvim-treesitter/assets/39000776/0f4c041e-a4d9-4eda-9434-24ae620c19b9)

Reply:
but precedence is not orthogonal to priority -- priority is (only!) needed to work around unwanted precedence. If you can ensure the wanted precedence directly, that is always preferable.

Reply:
I'm sorry I can't as I'm not familiar with the internal implementation.
Could you suggest me a change that works without priority?

Reply:
Just try moving the query to after the other query it should have priority over ("last highlight wins", and they are applied in the order they appear in this file).

Reply:
I've heard that and tried it. But it doesn't work as I said above.

To make the context clearer,

`(list a b c 1 2 3)` should be highlight as this

```racket
(list a b c 1 2 3)
(@variable @variable @variable @variable @number @number @number)
```

And `list` should also be highlight as a builtin function.

while when a quote wraps it, like this:

```racket
(quote (list a b c 1 2 3))
```

everything should be highlight as `@symbol`.

I think priority is necessary here.

I don't know why you don't like it so much though.

Reply:
> I don't know why you don't like it so much though.

1. Because it doesn't do what most people believe it does (the priority applies to the extmark used for highlighting, not for the query/capture itself; these still get all applied in their order!).2
2. Because it can interfere with injections.

So it's always worth trying to do something without priority first.

Reply:
Similarly, from a quick look at the queries, the `priority` is put in such a way that correlates to the order of highlight application itself, so it should not be needed if all of them were to be removed

Reply:
> so it should not be needed if all of them were to be removed

@lucario387 How? Give me a change suggestion to make it works under this context.

@clason I tried it many times and can't do it anyway as the nested level of list can be infinity.

You could

1. accept current implementation, or
2. accept another implementation which only highlight `'abc` and only single nested level list `'(list a b c)`. This implementation don't need priority.

If you accept case 2, then I will do it later.

Reply:
All I was asking was to try a different order first; I wasn't saying priorities are inacceptable.

Reply:
Then how? I tried it many times and can't implement. I just think it's impossible to implement without priority.

Reply:
First, as I said, priorities are ok if ordering is not enough; all we did was ask you to confirm that ordering is in fact not enough. You did, so that's fine with me. I don't understand why we keep discussing this -- increasingly acrimoniously...

But there may be a misunderstanding: the query that raised _our_ eyebrows is
```
(quote (#set! "priority" 105)) @symbol
```
(In fact, I suspect -- and hence our question -- that _all_ these priorities are not really necessary in nvim-treesitter, even though they _are_ necessary in `tree-sitter highlight`, which uses a different precedence rule ("first match wins", instead of "all matches are applied in order").